python/dolfinx/fem/function.py

```diff
@@ -746,11 +721,9 @@ def ufl_function_space(self) -> ufl.FunctionSpace:
         return self
 
     @property
-    def element(
-        self,
-    ) -> typing.Union[_cpp.fem.FiniteElement_float32, _cpp.fem.FiniteElement_float64]:
+    def element(self) -> FiniteElement:
         """Function space finite element."""
-        return self._cpp_object.element  # type: ignore
+        return FiniteElement(self._cpp_object.element)
```

Comment:
This create a new `FiniteElement` wrapper each time. Would it be better to wrap the element in the initialiser?

Reply:
That would require an additional state variable of the wrapper class `FunctionSpace` keeping track of the `FiniteElement`, which contradicts it being a wrapper to provide access only to the `cpp_object`, as we would then have two `finite_element`'s around to manage. 

Also the `FiniteElement` object itself is only a wrapper, it performs no computations or produces additional states, so it's fine to be created whenever we expose a `FiniteElement`.

Reply:
Perhaps this is naive, but could https://docs.python.org/3/library/functools.html#functools.cached_property be used?

Reply:
Seems to add exactly what we would need here - however I am not convinced we satisfy the
> Useful for expensive computed properties of instances that are otherwise effectively immutable.

part of it. Calling the constructor of `FiniteElement` is very lightweight and would then compete with the cached lookup.

Reply:
With non-cached approach calling the constructor each time, won't this break expected outcome of
```python
assert V.element is V.element
```
?

Reply:
No, the equal check is performed on the underlying `cpp_object` https://github.com/FEniCS/dolfinx/pull/3542/files#diff-93e5ad3a5a70702c7593dd13933baba17ded94b1c679cdc58129bd3e12610a3fR184

Reply:
However I did miss that in the other PR https://github.com/FEniCS/dolfinx/pull/3509.

Reply:
Ah right, `is` not the same as `==`, my bad. You are right, then this check will fail.

Reply:
Another aspect to consider regarding `cached_property`:
> Transform a method of a class into a property whose value is computed once and then cached as a normal attribute for the life of the instance.

In the current implementation, what is the lifespan of all these instanitated `FiniteElement`s?

Reply:
As I look through the Python layer, this question crops up a lot. Perhaps this can be addressed in a future PR.

Reply:
Sounds good, will file an issue when this gets merged.

Reply:
In `dolfinx.mesh.Mesh`, we store the Python-wrappers: https://github.com/FEniCS/dolfinx/blob/main/python/dolfinx/mesh.py#L275-L278 

I guess we need to decide on a unified structure for this.